app/src/main/java/com/nextcloud/talk/users/UserManager.kt

```diff
@@ -0,0 +1,433 @@
+                // unused atm
+            }
+
+            override fun onComplete() {
+                // unused atm
+            }
+        })
+
+        return Observable.just(result)
+    }
+
+    fun createOrUpdateUser(
+        username: String?,
+        userAttributes: UserAttributes,
+    ): Observable<UserNgEntity?> {
+        var user: UserNgEntity? = null
+
+        if (userAttributes.id == null && username != null && userAttributes.serverUrl != null) {
+            userRepository.getUserWithUsernameAndServer(username, userAttributes.serverUrl)
+                .subscribe(object : Observer<UserNgEntity?> {
+                    override fun onSubscribe(d: Disposable) {
+                        // unused atm
+                    }
+
+                    override fun onNext(userEntity: UserNgEntity) {
+                        user = userEntity
+                    }
+
+                    override fun onError(e: Throwable) {
+                        // unused atm
+                    }
+
+                    override fun onComplete() {
+                        // unused atm
+                    }
+                })
+        } else if (userAttributes.id != null) {
+            userRepository.getUserWithId(userAttributes.id)
+                .subscribe(object : Observer<UserNgEntity?> {
+                    override fun onSubscribe(d: Disposable) {
+                        // unused atm
+                    }
+
+                    override fun onNext(userEntity: UserNgEntity) {
+                        user = userEntity
+                    }
+
+                    override fun onError(e: Throwable) {
+                        // unused atm
+                    }
+
+                    override fun onComplete() {
+                        // unused atm
+                    }
+                })
+        }
+
+        if (user == null) {
+            user = createUser(
+                username,
+                userAttributes
+            )
+        } else {
+            updateUserData(
+                user!!,
+                userAttributes
+            )
+        }
+
+        userRepository.insertUser(user!!)
+        return userRepository.getUserWithIdLiveData(user!!.id)
+    }
+
+    fun getUserWithUsernameAndServer(username: String, server: String): Observable<UserNgEntity?> {
+        return userRepository.getUserWithUsernameAndServer(username, server)
+    }
+
+    private fun updateUserData(user: UserNgEntity, userAttributes: UserAttributes) {
+        updateUserIdIfNeeded(userAttributes, user)
+        updateTokenIfNeeded(userAttributes, user)
+        updateDisplayNameIfNeeded(userAttributes, user)
+        updatePushConfigurationStateIfNeeded(userAttributes, user)
+        updateCapabilitiesIfNeeded(userAttributes, user)
+        updateCertificateAliasIfNeeded(userAttributes, user)
+        updateExternalSignalingServerIfNeeded(userAttributes, user)
+        updateCurrentUserStatusIfNeeded(userAttributes, user)
+    }
```

Comment:
Besides the fact that this should take `User` instead of  the entity, all the `ifNeeded` methods seem superfluous and inefficient. Just do an `updateUser(user: User)` method and update always. If the data is the same as before then there will be no changes :shrug:

Reply:
In between (like yesterday) for once I experience NPEs due to the token being null for existing users that have been used before and lived on the device for some time. I haven't plugged in the storing of users via room yet, so I have no clue what might have happened killing of the tokens in the database...